### ISSUE TITLE
COMMUNITY: As a user, I would like to have different types of objects in a single data entry batch

### ISSUE THREAD

**Issue by ehrenfoss** [SalesforceFoundation/Cumulus#1354]
With some feedback from the NPSP sprint group in DC, I decided to explore the following feature.  (This is the first time I'm working on an open source project on GitHub, and the first time I've added anything to the NPSP since 2008, so please let me know if I'm making any mistakes here.  Thanks!)
1.  Allow multiple types of objects to coexist within a single batch.  E.g. a batch could contain 3 Leads, 2 Opportunities, and 5 Accounts.

**Comment by njjc** [comment]
Hi @ehrenfoss, this is great! We try to keep a single item per github issue for tracking purposes, but if you plan on working on these together this should be fine.

We've recently put out documentation on the technical parts of contributing code, which you can visit at http://developer.salesforcefoundation.org/Cumulus/Contributor/

I'm happy to help you through the process and answer any questions you have going forward. Let me know if you need anything!

**Comment by ehrenfoss** [comment]
Thanks @njjc!  Starting to dig into this. Is there documentation for creating the field sets that allow a user to add more objects to the batch tool?  I saw a screenshot in the appexchange listing but don't see that page in the current cumulus codebase (BDE_BatchEntrySettings, third screenshot)  https://appexchange.salesforce.com/listingDetail?listingId=a0N30000005txCAEAY

**Comment by njjc** [comment]
That's interesting, @ehrenfoss, I haven't actually used the non-NPSP version. Let me dig into what happened when we imported it and I'll get back to you on that part.

**Comment by ehrenfoss** [comment]
@njjc I suspect there was no need for a configuration page after the switch to field sets and so it was dropped.  If you find any other documentation for the NPSP version, let me know.  

BTW started to do a little coding over the weekend, got stuck on an Eclipse issue, but looks like that is resolved.  Meeting up with a few people at NTC in Austin in early March if any NPSP devs will be there?

**Comment by njjc** [comment]
So here's the deal! In the process of converting the batch data entry tool to work with NPSP, we removed the confusing settings pages and simplified the process, as you discovered in your third point. To get a custom object available for batch entry, three things need to be created:
- A lookup field on the custom object to the Batch__c object ([code here](https://github.com/SalesforceFoundation/Cumulus/blob/dev/src/classes/BDE_BatchDataEntry.cls#L127)).
- A field set on the custom object called "BDE_Entry_FS" containing the fields desired for entering data ([code here](https://github.com/SalesforceFoundation/Cumulus/blob/dev/src/classes/BDE_BatchDataEntry.cls#L88))
- A field set on the custom object with the API name "BDE_List_FS" containing the fields desired for viewing entered data ([code here](https://github.com/SalesforceFoundation/Cumulus/blob/dev/src/classes/BDE_BatchDataEntry.cls#L92))

So basically, we did away with the settings describing how to add new objects, but it appears we never documented the new process. This is definitely an area that we can use some help in. 

@jvarese, could you add creating a document for batch data entry to the backlog?

@ehrenfoss, are you interested in contributing this documentation?

Also, let me know if and when you'd like to set up a call to discuss any of this! Thanks again!

**Comment by njjc** [comment]
Yep, you beat me to it, but you're right on. Unfortunately we aren't going to NTC, but I'm happy to schedule a video chat any time.

**Comment by ehrenfoss** [comment]
@njjc would definitely be interested in contributing to documentation

**Comment by njjc** [comment]
Awesome! @jvarese will be happy to guide you in the process. Thanks so much, let me know if you need anything.

**Comment by jvarese** [comment]
@ehrenfoss - that's music to our ears. You can email me directly and we can chat about the doc. For reference, our growing doc set is housed here: https://powerofus.force.com/articles/Resource/Nonprofit-Starter-Pack-3-Documentation

**Comment by ehrenfoss** [comment]
I reached out to @jvarese to figure out documentation next steps. Split out the child object issue: https://github.com/SalesforceFoundation/Cumulus/issues/1413  I edited the issue title and description to narrow the focus here on support for multiple types of objects in a single batch.  

Next steps:

1.  Refactor initializeBatchEntry method and BDE_BatchDataEntry class to query multiple types of objects for a matching Batch ID.  
- Perhaps this can be done with a monster SOQL query with subselects for Batch child objects?  
- If not, might need to go beyond Batch__c.Object_Name__c field (Text 100) to provide for a list of object names for 1 query apiece.  
- Also need to keep track of fieldset / field list per object

2.  Refactor VisualForce page to nest a table within the record list - where the fields included in the list fieldset for each object will be shown.  Might be clunky / ugly if there's a big difference between fields included for each object.

3.  Test by editing a couple types of objects to have the same batch ID - can we view a list with multiple objects and edit each?  Does the saveBatch() method work?

4.  Make object selector accessible for in progress batches, and perhaps some buttons like "Add another <object name>" or "Add new object" 

5... ?

**Comment by ehrenfoss** [comment]
I refactored initializeBatchEntry to query multiple types of objects.  I created Batch__c.Included_Objects__c as a textarea (255) field, which should be OK for up to 10 object names at ~25 characters apiece.  This refactoring also sets up maps of objects and field list items, and BatchItems.

**Comment by njjc** [comment]
Awesome! Let me know if you want to review together at some point!

**Comment by ehrenfoss** [comment]
@njjc made some more progress today:  Finished refactoring to allow a list of objects of several types when displaying the record list.  

However, I'm now stuck.  

1.  I get Error: java.lang.UnsupportedOperationException when saving the VisualForce page including the following code. Probably due to [this error](https://developer.salesforce.com/forums/ForumsMain?id=906F00000009L6sIAE)
```
        <apex:outputPanel id="recLists">
        <apex:pageBlock mode="detail" rendered="{!itemCount>0}" >
            <apex:pageBlockSection columns="1" title="Saved Record Lists ({!itemCount} records)">
                <apex:repeat var="objname" value="{!objnames}" >
                <apex:pageBlockTable var="item" value="{!itemLists[objname]}" width="90%" >                   
                    <apex:column >
                        <apex:facet name="header">Action</apex:facet>
                        <apex:commandLink value="edit" action="{!item.editItem}" styleclass="editLink" immediate="true" rerender="recInfo" onclick="rememberDefaults();" oncomplete="resetSticky();" />
                    </apex:column>
                    <apex:repeat var="fieldname" value="{!listFieldLists[objname]}" > 
                        <apex:column value="{!item.sobj[fieldname]}" styleClass="{!fieldname}" />
                    </apex:repeat>
                </apex:pageBlockTable> 
                </apex:repeat>
            </apex:pageBlockSection>
        </apex:pageBlock>
        </apex:outputPanel>
```

2.  After removing that, I'm now getting a Insufficient Privileges error message on the VF page.  I remember hitting these years ago and they are very, very annoying.  Tests pass, I've tried executing the controller and the initialize method in anonymous execution...no dice.

I think I need help!  Would be a good opportunity to walk you through my overall approach to make sure it's within bounds.

**Comment by njjc** [comment]
Hi @ehrenfoss, great to hear about the progress. I'm happy to spend some time to go over the project with you, how about some time on Wednesday or Thursday of this week? Shoot me an email at ncampbell@salesforce.com

**Comment by ehrenfoss** [comment]
@njjc I think I have everything going smoothly except for the list fieldsets.  Check it out and let me know what you think: https://www.youtube.com/watch?v=xI5ayPAm7cE

Should I do a pull request?

**Comment by ehrenfoss** [comment]
Pushed to branch ehrenfoss-issue-1354.  Still have a bug with the "new object" button and some unresolved issues w/ item list.  Haven't tested thoroughly either.

**Comment by kbromer** [comment]
**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x800000001fuAAAQ" target="_blank">W-008293</a>

**Comment by njjc** [comment]
This has been marked as "Never" internally.
